File: Halving.py
def check_valid_pair(a, b, rule, target):
    if rule == 0:  # min rule
        return min(a, b) == target
    else:  # max rule
        return max(a, b) == target

def solve(N, C, R, B):
    MOD = 998244353
    
    # Step 1: Find known numbers in C
    known = {x for x in C if x != -1}
    # Step 2: Create list of available numbers
    available = [x for x in range(1, 2*N + 1) if x not in known]
    
    def find_valid_arrangements(pos, used, C_copy):
        if pos >= N:
            return 1
            
        count = 0
        pos1, pos2 = 2*pos, 2*pos + 1
        target = B[pos]
        rule = R[pos]
        
        # Case 1: Both positions are known
        if C_copy[pos1] != -1 and C_copy[pos2] != -1:
            if check_valid_pair(C_copy[pos1], C_copy[pos2], rule, target):
                return find_valid_arrangements(pos + 1, used, C_copy)
            return 0
            
        # Case 2: One position is known
        elif C_copy[pos1] != -1:
            known_val = C_copy[pos1]
            for val in available:
                if val not in used and check_valid_pair(known_val, val, rule, target):
                    used.add(val)
                    C_copy[pos2] = val
                    count = (count + find_valid_arrangements(pos + 1, used, C_copy)) % MOD
                    used.remove(val)
                    C_copy[pos2] = -1
        elif C_copy[pos2] != -1:
            known_val = C_copy[pos2]
            for val in available:
                if val not in used and check_valid_pair(val, known_val, rule, target):
                    used.add(val)
                    C_copy[pos1] = val
                    count = (count + find_valid_arrangements(pos + 1, used, C_copy)) % MOD
                    used.remove(val)
                    C_copy[pos1] = -1
                    
        # Case 3: Both positions are unknown
        else:
            for i in range(len(available)):
                val1 = available[i]
                if val1 in used:
                    continue
                for j in range(i + 1, len(available)):
                    val2 = available[j]
                    if val2 in used:
                        continue
                    if check_valid_pair(val1, val2, rule, target):
                        used.add(val1)
                        used.add(val2)
                        C_copy[pos1] = val1
                        C_copy[pos2] = val2
                        count = (count + find_valid_arrangements(pos + 1, used, C_copy)) % MOD
                        used.remove(val1)
                        used.remove(val2)
                        C_copy[pos1] = -1
                        C_copy[pos2] = -1
                        
        return count
    
    return find_valid_arrangements(0, set(), C[:])

def main():
    N = int(input())
    C = list(map(int, input().split()))
    R = list(map(int, input().split()))
    B = list(map(int, input().split()))
    
    print(solve(N, C, R, B))

if __name__ == "__main__":
    main()

File: Halving.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>

using namespace std;

const int MOD = 998244353;

int factorial(int n) {
    long long result = 1;
    for (int i = 2; i <= n; ++i) {
        result = (result * i) % MOD;
    }
    return result;
}

int main() {
    int N;
    cin >> N;

    vector<int> C(2 * N);
    for (int i = 0; i < 2 * N; ++i) {
        cin >> C[i];
    }

    vector<int> R(N);
    for (int i = 0; i < N; ++i) {
        cin >> R[i];
    }

    vector<int> B(N);
    for (int i = 0; i < N; ++i) {
        cin >> B[i];
    }

    // Step 1: Identify the used numbers from C and B
    set<int> used;
    for (int i = 0; i < 2 * N; ++i) {
        if (C[i] != -1) {
            used.insert(C[i]);
        }
    }
    for (int i = 0; i < N; ++i) {
        used.insert(B[i]);
    }

    // Step 2: Identify missing numbers
    vector<int> missing;
    for (int i = 1; i <= 2 * N; ++i) {
        if (used.find(i) == used.end()) {
            missing.push_back(i);
        }
    }

    // Step 3: Check the relationships based on R and B
    bool valid = true;
    vector<pair<int, int>> unknownPairs; // To store pairs of unknowns
    for (int i = 0; i < N; ++i) {
        int a1 = C[2 * i], a2 = C[2 * i + 1];

        // If both are unknown, we add them to unknownPairs
        if (a1 == -1 && a2 == -1) {
            unknownPairs.push_back({i, i}); // Mark this index
            continue;
        }

        // Assign known values if one is unknown
        if (a1 == -1) {
            a1 = (R[i] == 0) ? B[i] : -1; // Min case
        } else if (a2 == -1) {
            a2 = (R[i] == 0) ? B[i] : -1; // Min case
        }

        // Now check if the known values satisfy the conditions
        if (R[i] == 0) { // min case
            if (min(a1, a2) != B[i]) {
                valid = false;
                break;
            }
        } else { // max case
            if (max(a1, a2) != B[i]) {
                valid = false;
                break;
            }
        }
    }

    // Handle unknown pairs separately
    if (valid) {
        int numMissing = missing.size();
        int count = 1;

        for (const auto& pair : unknownPairs) {
            int index = pair.first;
            if (R[index] == 0) {
                // Min case: both must be the same value
                if (numMissing < 1) {
                    valid = false; // No missing numbers to fill
                    break;
                } else {
                    // Choose one missing number for both unknowns
                    count = (count * numMissing) % MOD;
                    numMissing--; // Reduce the count of missing for the next pair
                }
            } else {
                // Max case: they must be different
                if (numMissing < 2) {
                    valid = false; // Not enough missing numbers
                    break;
                } else {
                    count = (count * numMissing * (numMissing - 1)) % MOD; // Choose two distinct numbers
                    numMissing -= 2; // Reduce for the next pair
                }
            }
        }

        // Step 4: If valid, output the count
        if (valid) {
            cout << count << endl;
        } else {
            cout << 0 << endl;
        }
    } else {
        cout << 0 << endl;
    }

    return 0;
}
